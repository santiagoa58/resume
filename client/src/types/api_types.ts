/**
 * RESUME API TYPES
 */

// resume metadata
export interface IResumeMetadata {
  // resume file name
  name: string;
  // resume file id
  id: string;
}

// resume work experience
export interface IResumeExperience {
  // name of the company
  company: string;
  // location of the company
  location: string;
  // duration of the work experience
  duration: string;
  // role of the resume author
  role: string;
  // list of responsibilities
  responsibilities: string[];
}

// resume education
export interface IResumeEducation {
  // degree obtained by the resume author
  degree: string;
  // institution where the resume author obtained their degree
  institution: string;
  // duration of the education
  duration: string;
}

// resume personal project
export interface IResumePersonalProject {
  // name of the personal project
  name: string;
  // description of the personal project
  description: string;
}

// Resume details
export interface IResume {
  // id of the resume
  doc_id: string;
  // name of the resume
  doc_name: string;
  // name of the resume author
  name: string;
  // list of contact information for the resume author
  contacts: string[];
  // location of the resume author
  location: string;
  // professional title of the resume author
  title: string;
  // summary section of the resume
  summary: string;
  // list of skills
  skills: string[];
  // list of work experience
  experiences: IResumeExperience[];
  // education section of the resume
  educations: IResumeEducation[];
  // list of personal projects mentioned in the resume
  personal_projects: IResumePersonalProject[];
}

/**
 * PROJECTS API TYPES
 */

// A GitHub user.
interface IProjectUser {
  name?: string | null;
  avatar_url: string;
  gravatar_id?: string | null;
  html_url: string;
  type: string;
}

export interface IProject {
  id: number;
  name: string;
  full_name: string;
  owner: IProjectUser;
  private: boolean;
  html_url: string;
  description: string | null;
  fork: boolean;
  url: string;
  languages_url: string;
  homepage: string | null;
  language: string | null;
  forks_count: number;
  stargazers_count: number;
  watchers_count: number;
  /**
   * The size of the repository. Size is calculated hourly. When a repository is initially created, the size is 0.
   */
  size: number;
  default_branch: string;
  topics: string[];
  archived: boolean;
  disabled: boolean;
  visibility: string;
  pushed_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  subscribers_count?: number;
  forks: number;
  open_issues: number;
  watchers: number;
  languages: string[];
}
