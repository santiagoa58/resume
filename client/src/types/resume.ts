export type ResumeRole =
  | 'frontend'
  | 'backend'
  | 'fullstack'
  | 'devops'
  | 'qa'
  | 'pm'
  | 'designer'
  | 'machine-learning'
  | 'data-scientist'
  | 'security'
  | 'other';
