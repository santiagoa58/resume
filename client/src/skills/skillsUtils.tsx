import React, { FC } from 'react';
import Typography from '@mui/material/Typography';
import { IProject } from '../types/api_types';
import { getUniqueArrayEntries } from '../utils/objectUtils';
import {
  MAIN_PROJECT,
  OUTDATED_PROJECT,
  SKILLS_VARIETY,
} from '../utils/constants';

export enum COMMANDS {
  LIST_SKILLS = 'list skills',
  HELP = 'help',
  CLEAR = 'clear',
  LIST_PROJECT_SKILLS = 'list project skills',
  // command to describe a project
  DESCRIBE_PROJECT = 'describe',
  LIST_PROJECTS = 'list projects',
  SKILL_SEARCH = 'search',
}

// map of all commands as key and descriptions as value
export const COMMANDS_DESCRIPTION: Record<COMMANDS, string> = {
  [COMMANDS.LIST_SKILLS]: 'List all my skills',
  [COMMANDS.LIST_PROJECTS]: 'List all my projects',
  [COMMANDS.HELP]: 'List all available commands',
  [COMMANDS.CLEAR]: 'Clear the terminal',
  [COMMANDS.LIST_PROJECT_SKILLS]: 'List skills mentioned in each project',
  [COMMANDS.SKILL_SEARCH]:
    "Search for the mention of a skill. ex: 'search react'",
  [COMMANDS.DESCRIBE_PROJECT]:
    "gets the project description for a given project name. If no project is provided this website will be described. ex: 'describe projectName' | 'describe'",
};

export const HighlightedSpan: FC<{ value: string }> = ({ value }) => (
  <Typography component="span" sx={{ color: 'info.main' }}>
    {value}
  </Typography>
);

/**
 * tag function to create a Typography component with highlighted values
 * #### EXAMPLE:
 ```
commandOutputTagFunction`Type ${COMMAND.LIST_SKILLS} to view a list of all my skills.`;
 ```
 * #### RETURNS:
 ```
<Typography>
    Type <Typography color="success">list skills</Typography> to view a list of all my skills.
</Typography>
 ```
 * */
export const commandOutputTagFunction = (
  strings: TemplateStringsArray,
  ...values: string[]
) => {
  return (
    <Typography paragraph>
      {strings.map((string, index) => {
        // get the value if it exists
        const value = values.length > index ? values[index] : undefined;
        return (
          <React.Fragment key={`string-and-value-${string}-${value}-${index}`}>
            {string} {value !== undefined && <HighlightedSpan value={value} />}
          </React.Fragment>
        );
      })}
    </Typography>
  );
};

export const normalizeCommand = (command: string) => {
  return command.toLowerCase().trim();
};

const normalizeSkill = (skill: string): string => {
  // replace all _ or - with a space
  return skill.replace(/[_-]/g, ' ').toLowerCase().trim();
};

const isValidSkill = (normalizedSkill: string): boolean => {
  if (
    normalizedSkill === OUTDATED_PROJECT ||
    normalizedSkill === MAIN_PROJECT ||
    !normalizedSkill
  ) {
    return false;
  }
  return true;
};

const normalizeSkillWithConsistentName = (topic: string): string => {
  const normalizedTopic = normalizeSkill(topic)
    .split(' ')
    .map(getConsistentSkillName)
    .join(' ');
  return getConsistentSkillName(normalizedTopic);
};

export const getProjectSkills = (projects: IProject[]) => {
  return projects.map((project) => {
    const normalizedLanguages = project.languages.map(
      normalizeSkillWithConsistentName
    );
    const normalizedTopics = project.topics.map(
      normalizeSkillWithConsistentName
    );
    const skillSet = getUniqueArrayEntries(
      normalizedLanguages.concat(normalizedTopics).filter(isValidSkill),
      (skill, uniqueSkillSet) => {
        const alreadyExistingSkill = skill
          .split(' ')
          .find((skillPart) => uniqueSkillSet.has(skillPart));
        if (alreadyExistingSkill) {
          // value is not unique
          return [false, alreadyExistingSkill];
        }
        // value is unique
        return true;
      },
      // set to false to keep the last value
      false
    );
    return {
      projectName: project.name,
      skillSet,
    };
  });
};

export const WELCOME_MESSAGE: JSX.Element = commandOutputTagFunction`Welcome to the Skills Terminal! To view a list of all my skills type ${COMMANDS.LIST_SKILLS}. To view all the available commands enter ${COMMANDS.HELP}.`;

export const getConsistentSkillName = (skill: string): string => {
  const lowerCaseSkill = skill.toLowerCase().trim();
  for (const [consistentSkillName, skillVariety] of Object.entries(
    SKILLS_VARIETY
  )) {
    if (skillVariety.has(lowerCaseSkill)) {
      return consistentSkillName.toLowerCase();
    }
  }
  return skill.toLowerCase();
};
