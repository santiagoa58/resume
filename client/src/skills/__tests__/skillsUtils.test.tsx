// test skillsUtils
import { IProject } from '../../types/api_types';
import * as utils from '../skillsUtils';

test('getConsistentSkillName', () => {
  expect(utils.getConsistentSkillName('react')).toEqual('react');
  expect(utils.getConsistentSkillName('reactjs')).toEqual('react');
  expect(utils.getConsistentSkillName('react.js')).toEqual('react');
  expect(utils.getConsistentSkillName('frontend')).toEqual('frontend');
  expect(utils.getConsistentSkillName('front-end')).toEqual('frontend');
  expect(utils.getConsistentSkillName('front end')).toEqual('frontend');
  expect(utils.getConsistentSkillName('front_end')).toEqual('frontend');
  expect(utils.getConsistentSkillName('backend')).toEqual('backend');
  expect(utils.getConsistentSkillName('back-end')).toEqual('backend');
  expect(utils.getConsistentSkillName('back end')).toEqual('backend');
  expect(utils.getConsistentSkillName('back_end')).toEqual('backend');
  expect(utils.getConsistentSkillName('fullstack')).toEqual('fullstack');
  expect(utils.getConsistentSkillName('full-stack')).toEqual('fullstack');
  expect(utils.getConsistentSkillName('full stack')).toEqual('fullstack');
  expect(utils.getConsistentSkillName('full_stack')).toEqual('fullstack');
  expect(utils.getConsistentSkillName('javascript')).toEqual('javascript');
  expect(utils.getConsistentSkillName('js')).toEqual('javascript');
  expect(utils.getConsistentSkillName('typescript')).toEqual('typescript');
  expect(utils.getConsistentSkillName('ts')).toEqual('typescript');
  expect(utils.getConsistentSkillName('python')).toEqual('python');
  expect(utils.getConsistentSkillName('py')).toEqual('python');
  expect(utils.getConsistentSkillName('aws')).toEqual('aws');
  expect(utils.getConsistentSkillName('amazon web services')).toEqual('aws');
});
describe('commandOutputTagFunction', () => {
  it('should return the correct JSX', () => {
    const result = utils.commandOutputTagFunction`Type ${utils.COMMANDS.LIST_SKILLS} to view a list of all my skills.`;
    expect(result).toMatchSnapshot();
  });
});
describe('normalizeCommand', () => {
  it('should return the correct value', () => {
    expect(utils.normalizeCommand('list skills')).toEqual(
      utils.COMMANDS.LIST_SKILLS
    );
    expect(utils.normalizeCommand('list skills ')).toEqual(
      utils.COMMANDS.LIST_SKILLS
    );
    expect(utils.normalizeCommand(' list skills')).toEqual(
      utils.COMMANDS.LIST_SKILLS
    );
    expect(utils.normalizeCommand(' list skills ')).toEqual(
      utils.COMMANDS.LIST_SKILLS
    );
  });
});
describe('getProjectSkills', () => {
  it('should return the correct value', () => {
    const projects: Partial<IProject>[] = [
      {
        name: 'project 1',
        description: 'project 1 description',
        languages: ['javascript', 'ts'],
        topics: ['aws', 'css', 'VANILLA JS'],
      },
      {
        name: 'project 2',
        description: 'project 2 description',
        languages: ['python', 'ts', 'css'],
        topics: ['front-end', 'back end', 'advanced-css-and-sass'],
      },
    ];
    const result = utils.getProjectSkills(projects as IProject[]);
    expect(result).toEqual([
      {
        projectName: 'project 1',
        skillSet: ['typescript', 'aws', 'css', 'vanilla javascript'],
      },
      {
        projectName: 'project 2',
        skillSet: [
          'python',
          'typescript',
          'frontend',
          'backend',
          'advanced css and sass',
        ],
      },
    ]);
  });
});
