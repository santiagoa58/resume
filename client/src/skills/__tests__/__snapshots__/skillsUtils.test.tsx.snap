// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commandOutputTagFunction should return the correct JSX 1`] = `
<ForwardRef(Typography)
  paragraph={true}
>
  <React.Fragment>
    Type 
     
    <HighlightedSpan
      value="list skills"
    />
  </React.Fragment>
  <React.Fragment>
     to view a list of all my skills.
     
  </React.Fragment>
</ForwardRef(Typography)>
`;
